src/script/lua_api/l_object.cpp

```diff
@@ -684,6 +684,85 @@ int ObjectRef::l_get_properties(lua_State *L)
 	return 1;
 }
 
+// set_observers(self, observers)
+int ObjectRef::l_set_observers(lua_State *L)
+{
+	GET_ENV_PTR;
+	ObjectRef *ref = checkObject<ObjectRef>(L, 1);
+	ServerActiveObject *sao = getobject(ref);
+	if (sao == nullptr)
+		throw LuaError("Invalid ObjectRef");
+
+	// Reset object to "unmanaged" (sent to everyone)?
+	if (lua_isnoneornil(L, 2)) {
+		ServerActiveObject *parent = sao->getParent();
+		if (parent != nullptr && !parent->m_observer_names.has_value())
+			throw LuaError("Child observers need to be managed if parent observers are managed");
+		sao->m_observer_names.reset();
+		return 0;
+	}
+
+	std::unordered_set<std::string> observer_names;
+	lua_pushnil(L);
+	while (lua_next(L, 2) != 0) {
+		observer_names.insert(readParam<std::string>(L, -2));
+		if (!lua_toboolean(L, -1)) // falsy value?
+			throw LuaError("Values in the `observers` table need to be truthy");
+		lua_pop(L, 1); // pop value, keep key
+	}
+
+	RemotePlayer *player = getplayer(ref);
+	if (player != nullptr) {
+		if (observer_names.find(player->getName()) == observer_names.end())
+			throw LuaError("Players need to observe themselves");
+	}
+
+	// Check attachments
+	ServerActiveObject *parent = sao->getParent();
+	if (parent != nullptr) {
+		for (auto name : observer_names) {
+			if (!parent->isObservedBy(name))
+				throw LuaError("Child observers not a subset of parent observers");
+		}
+	}
+	const std::unordered_set<int> child_ids = sao->getAttachmentChildIds();
+	for (auto child_id : child_ids) {
+		ServerActiveObject *child = env->getActiveObject(child_id);
+		if (!child->m_observer_names.has_value())
+			throw LuaError("Child observers need to be managed if parent observers are managed");
+		for (auto name : child->m_observer_names.value()) {
+			if (!sao->isObservedBy(name))
+				throw LuaError("Child observers not a subset of parent observers");
+		}
+	}
+	sao->m_observer_names = observer_names;
+	return 0;
+}
+
+// get_observers(self)
+int ObjectRef::l_get_observers(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkObject<ObjectRef>(L, 1);
+	ServerActiveObject *sao = getobject(ref);
+	if (sao == nullptr)
+		throw LuaError("invalid ObjectRef");
+
+	const auto &observer_names = sao->m_observer_names;
+	if (!observer_names.has_value()) {
+		lua_pushnil(L);
+		return 1;
+	}
```

Comment:
this is more elegant imo:
```suggestion
	if (!sao->m_observer_names.has_value()) {
		lua_pushnil(L);
		return 1;
	}
	const auto &observer_names = sao->m_observer_names.value();
```